src/formats/TRR.cpp

```diff
@@ -0,0 +1,117 @@
+// Chemfiles, a modern library for chemistry file reading and writing
+// Copyright (C) Guillaume Fraux and contributors -- BSD license
+
+#include "chemfiles/formats/TRR.hpp"
+#include "chemfiles/files/TRRFile.hpp"
+
+#include "chemfiles/ErrorFmt.hpp"
+#include "chemfiles/Frame.hpp"
+using namespace chemfiles;
+
+template <> FormatInfo chemfiles::format_information<TRRFormat>() {
+    return FormatInfo("TRR").with_extension(".trr").description("TRR binary format");
+}
+
+#define STRING_0(x) #x
+#define STRING(x) STRING_0(x)
+#define CHECK(x) check_trr_error((x), (STRING(x)))
+
+TRRFormat::TRRFormat(std::string path, File::Mode mode, File::Compression compression)
+    : trr_(std::move(path), mode) {
+    if (compression != File::DEFAULT) {
+        throw format_error("TRR format does not support compression");
+    }
+}
+
+size_t TRRFormat::nsteps() { return static_cast<size_t>(trr_.nframes()); }
+
+void TRRFormat::read_step(size_t step, Frame& frame) {
+    step_ = static_cast<int64_t>(step);
+    CHECK(xdr_seek(trr_, trr_.offset(step_), SEEK_SET));
+    read(frame);
+}
+
+void TRRFormat::read(Frame& frame) {
+    int natoms = trr_.natoms();
+    int md_step = 0;
+    float time = 0;
+    float lambda = 0;
+    matrix box;
+    std::vector<float> x(static_cast<size_t>(natoms) * 3);
+    std::vector<float> v(static_cast<size_t>(natoms) * 3);
+    std::vector<float> f(static_cast<size_t>(natoms) * 3);
+    uint8_t has_prop = 0;
+    bool has_box = false;
+    bool has_positions = false;
+    bool has_velocities = false;
+
+    CHECK(read_trr(trr_, natoms, &md_step, &time, &lambda, box,
+                   reinterpret_cast<float(*)[3]>(x.data()), reinterpret_cast<float(*)[3]>(v.data()),
+                   NULL, &has_prop));
+
+    has_box = bool(has_prop & TRR_HAS_BOX);
+    has_positions = bool(has_prop & TRR_HAS_POSITIONS);
+    has_velocities = bool(has_prop & TRR_HAS_VELOCITIES);
+
+    frame.set("md_step", md_step); // actual step of MD Simulation
```

Comment:
You can use `frame.set_step` for this, since this is the intended usage of it =). The same applies to XTC

Reply:
Done for both, was confused, because I thought step in chemfiles was counted differently.

Reply:
>  was confused, because I thought step in chemfiles was counted differently.